File: Crio/crio_programming_interview_problems-master/InorderSuccessor/InorderSuccessor.cpp
#include <bits/stdc++.h>
#include "../crio/cpp/ds/TreeNode/TreeNode.hpp"
#include "Solution.cpp"

using namespace std;
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
int main()
{
    vector<TreeNode* > nodes = TreeNode().readTreeReturnAllNodes();
    TreeNode* root = nodes[0];
    assert(TreeNode().isValidateBST(root) == true);
    int queries;
    cin>>queries;
    while(queries--)
    {
        int index;
        cin>>index;
        TreeNode* temp = nodes[index];
        TreeNode* succ =  Solution().inOrderSuccessor(root, temp); 
        if(succ !=  NULL) 
            cout<<succ->val<<"\n";
        else
            cout<<"-1\n";
    }
    return 0;
}

File: Crio/crio_programming_interview_problems-master/InorderSuccessor/InorderSuccessor.java
import java.util.*;
import crio.ds.Tree.TreeNode;
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
class InorderSuccessor{
	public static void main(String[] args) {
		Scanner sc=new Scanner(System.in);
		Vector<TreeNode> nodes=TreeNode.readTreeReturnAllNodes(sc);
		int q=sc.nextInt();
		for(int i=0;i<q;i++)
		{	
			int index=sc.nextInt();
			Solution so=new Solution();
			System.out.println(so.inorderSuccessor(nodes.get(0),nodes.get(index)));
		}
	}
}